File: kyonggiyo-infrastructure/src/main/java/kyonggiyo/persistence/image/ImageBatchRepositoryImpl.java
package kyonggiyo.persistence.image;

import kyonggiyo.domain.image.Image;
import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.jdbc.core.namedparam.SqlParameterSource;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
@RequiredArgsConstructor
public class ImageBatchRepositoryImpl implements ImageBatchRepository {

    private final NamedParameterJdbcTemplate jdbcTemplate;

    @Override
    public void saveAllInBatch(List<Image> images) {
        SqlParameterSource[] sqlParameterSources = new SqlParameterSource[images.size()];
        int idx = 0;

        for (Image image : images) {
            SqlParameterSource source = new MapSqlParameterSource()
                    .addValue("upload_key", image.getKey())
                    .addValue("image_type", image.getImageType().name())
                    .addValue("reference_id", image.getReferenceId())
                    .addValue("created_at", image.getCreatedAt())
                    .addValue("updated_at", image.getUpdatedAt());
            sqlParameterSources[idx++] = source;
        }
        String sql = """
                insert into images (upload_key, image_type, reference_id, created_at, updated_at)
                values(:upload_key, :image_type, :reference_id, :created_at, :updated_at)""";
        jdbcTemplate.batchUpdate(sql, sqlParameterSources);
    }

}

File: kyonggiyo-infrastructure/src/main/java/kyonggiyo/image/repository/ImageBatchRepositoryImpl.kt
package kyonggiyo.image.repository

import kyonggiyo.application.image.domain.entity.Image
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.core.namedparam.SqlParameterSource
import org.springframework.stereotype.Repository

@Repository
class ImageBatchRepositoryImpl(
        private val jdbcTemplate: NamedParameterJdbcTemplate
) : ImageBatchRepository {

    override fun saveAllInBatch(images: List<Image>) {
        val sqlParameterSources = arrayOfNulls<SqlParameterSource>(images.size)
        for ((idx, image) in images.withIndex()) {
            val source: SqlParameterSource = MapSqlParameterSource()
                    .addValue("upload_key", image.key)
                    .addValue("image_type", image.imageType.name)
                    .addValue("reference_id", image.referenceId)
                    .addValue("created_at", image.createdAt)
                    .addValue("updated_at", image.updatedAt)
            sqlParameterSources[idx] = source
        }
        val sql = """
                insert into images (upload_key, image_type, reference_id, created_at, updated_at)
                values(:upload_key, :image_type, :reference_id, :created_at, :updated_at)
                """
        jdbcTemplate.batchUpdate(sql, sqlParameterSources)
    }

}
